source/main/physics/Beam.cpp

```diff
@@ -2568,7 +2568,7 @@ void Actor::calcShocks2(int beam_i, Real difftoBeamL, Real& k, Real& d, Real dt,
                     ar_beams[i].shock->trigger_switch_state = 0.0f;
                 if (ar_beams[i].shock->flags & SHOCK_FLAG_TRG_BLOCKER) // this is an enabled blocker and past boundary
                 {
-                    for (int scount = i + 1; scount <= i + ar_beams[i].shock->trigger_cmdshort; scount++) // (cycle blockerbeamID +1) to (blockerbeamID + beams to lock)
+                    for (int scount = i + 1; scount <= std::min(i + ar_beams[i].shock->trigger_cmdshort, ar_num_beams - 1); scount++) // (cycle blockerbeamID +1) to (blockerbeamID + beams to lock)

```

Comment:
The  added `min()` is a hack which hasn't been here before: https://github.com/RigsOfRods/rigs-of-rods/blame/b45238e29512858b4f8e487c127698d8fa31b5d3/source/main/physics/Beam.cpp#L2606

This is what I had in mind when I talked about introducing subtle trigger bugs

Reply:
`ar_beams` used to be an array with a fixed number of elements. Hence no crash if you tried to access a beam beyond `ar_num_beams`.

`ar_beams[scount].shock` used to return false in those cases now it just crashes.

Using `std::min` is equivalent to `ar_beams[scount].shock == false`. The if block is not executed.

Reply:
I'm positive the `ar_beams[scount].shock != nullptr` check doesn't mean "Am I still within beam count?" but simply "is this beam a shock?". 

RoR's physics are all done in means of "loop everything, switch over item type and process" and in this case, we're only filtering out shocks from all the beams. An optimal solution would be to have a separate array of `shock_t` with added beam index - then we could simply iterate over that. Also, separating actual shock absorbers from triggers would be nice.

Reply:
> was never intended to check "Am I still within beam count?"

We need to replicate the old behavior nonetheless.

> An optimal solution

I'm really only looking for a proper fix right now. We have way to many pending issues to even think about refactoring `calcShocks`.

> loop everything

If the truck file is setup correctly then this loop only loops over shocks.
The if check is only there as a "failsafe".